Filter aggregator: add ne (not equal) and make it work on string data

brunoballekens opened the pull request:
Added `ne` (not equal) operator 
Changes to make the aggregator work on string data

(sorry for all whitespace changes)

merged upstream by https://github.com/kairosdb/kairosdb/pull/582

Commit:
Filter aggregator: add ne (not equal) and make it work on string data

gabrielreid commented on src/main/java/org/kairosdb/core/aggregator/FilterAggregator.java:
```diff
@@ -19,121 +19,126 @@
+            boolean foundValidDp = false;
+            while (!foundValidDp && currentDataPoint != null) {
+                boolean filterOutValue = false;
+                if (Double.class.isAssignableFrom(m_threshold.getClass())) {
```
If the threshold value is a Long or an Integer (assuming that that is possible), then string comparison will be used here.

I think that changing this to the following will resolve this.
```
if (Number.class.isAssignableFrom(m_threshold.getClass())) {
    filterOutValue = filterOut(m_filterop, currentDataPoint.getDoubleValue(), ((Number) m_threshold).doubleValue());
}
```

brunoballekens replied:
Changed

gabrielreid commented on src/main/java/org/kairosdb/core/aggregator/FilterAggregator.java:
```diff
@@ -19,121 +19,126 @@
+        public boolean hasNext() {
+            boolean foundValidDp = false;
+            while (!foundValidDp && currentDataPoint != null) {
+                boolean filterOutValue = false;
```
It's probably a better to default `filterOutValue` to true instead of false.

As it currently stands, the filter condition will only be checked if the threshold is numeric (actually, if it's a double) *or* if the the data point value is a string.

If the threshold is a string, and the data point value is numeric, then then the data point will never be filtered out. I think that in case of a mismatch between the type of the threshold and the type of the data point value we always want to filter out the data point.

gabrielreid reviewed (changes requested):
It looks like there are a few edge cases with combinations of types that will give some problems.

Commit:
Filter aggregator: add ne (not equal) and make it work on string data
PR comments

gabrielreid commented on src/main/java/org/kairosdb/core/aggregator/FilterAggregator.java:
```diff
@@ -109,14 +109,25 @@ public FilterDataPointAggregator(DataPointGroup innerDataPointGroup) {
+                } else {
+                    // when types are different, we filter consider the datapoint to be non-equal to the threshold, so we just check if the operation is non-equal
+                    // for all other operations the datapoint will NOT be filtered out
+                    filterOutValue = FilterOperation.NE.equals(m_filterop);
```
It's probably not such an important edge case, but I think that we might want to invert this, i.e. make it
```
filterOutValue = !FilterOperation.EQUAL.equals(m_filterop);
```

As it currently stands, doing a `GT` filter on numerical data points with a string threshold (or the same with string data points and numerical threshold) will allow all records through. I think that the general intention of e.g. a `GT` filter is "give me all values `x` where `(x > threshold) == true`". If x is a string and threshold is a double, then `x > threshold` can never true, so `x` should be dropped, but right now that `x` will be included in the returned results.

This also translates to inverting this filter downstream, so if we say that we want all data points where the `x > 100`, then we'll translate that into a filter of `x <= 100`, which would currently give back all numerical values that are less than equal to 100, as well as giving back all string values.

brunoballekens replied:
Changed the code: if dataPoint type and threshold type are different, the datapoint *WILL* be filtered out

gabrielreid reviewed (approved):
The cross-type filtering might still need a change (see comment).

Commit:
Filter aggregator: add ne (not equal) and make it work on string data
If types (datapoint type vs threshold type) are different, data will be filtered out

francisdb reviewed (commented):
No test cases? Same comment about reverting whitespace changes

Commit:
Filter aggregator: add ne (not equal) and make it work on string data
Added unit tests
